src/NServiceBus.AzureStoragePersistence/CloudTableExtensions.cs

```diff
@@ -16,9 +18,18 @@ public static class CloudTableExtensions
             {
                 var seg = await table.ExecuteQuerySegmentedAsync(query, token, ct).ConfigureAwait(false);
                 token = seg.ContinuationToken;
-                items.AddRange(seg);
 
-            } while (token != null && !ct.IsCancellationRequested);
+                if (items.Count + seg.Count() > take)
```

Comment:
If a collection behind it does not support counting, seg.Count will be iterating items over and over. Maybe variable extraction would be better?